### PR TITLE
Feature/vertex label namespaces in python

### PR DESCRIPTION
**Pull request by migwellian**

### CODE REVIEW

File: keanu-python/keanu/base.py [hunk 1]
```diff
@@ -13,9 +13,14 @@ def __repr__(self) -> str:
...
 
     def __getattr__(self, k: str) -> Callable:
+        self.__check_if_constructed_without_error(k)
         self.__check_if_unwrapped(k)
         raise AttributeError("{} has no attribute {}".format(self.__class__, k))
 
+    def __check_if_constructed_without_error(self, k: str) -> None:
+        if k in ("_val", "_class"):
+            raise ValueError("Object did not get properly constructed - probably due to an earlier unhandled Error.")
+
```
Comment (migwellian) [inline keanu-python/keanu/base.py]: this is unrelated to this PR, but it sometimes bites me - typically when a ctor throws an exception, and in the exception there is some kind of error string that ends up calling `__getattr___`

File: keanu-python/tests/test_base.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ def test_you_can_call_a_java_method_on_the_unwrapped_object(java_list_wrapper) -
 
 
 def test_you_cannot_call_a_java_method_with_snake_case_on_the_unwrapped_object(java_list_wrapper) -> None:
-    with pytest.raises(Py4JError, match="Method is_empty\(\[\]\) does not exist"):
+    with pytest.raises(Py4JError, match=r"Method is_empty\(\[\]\) does not exist"):
```
Comment (migwellian) [inline keanu-python/tests/test_base.py]: unrelated to this PR - just to remove a warning in the tests

File: keanu-python/tests/test_labels.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+def test_there_is_a_helper_method_to_build_the_string_from_a_list() -> None:
+    label = _VertexLabel.create_from_list("outer", "inner", "foo")
+    assert label.unwrap().getUnqualifiedName() == "foo"
+    assert label.get_name() == "outer.inner.foo"
+
+
+def test_the_repr_method_makes_it_clear_what_the_namespace_is() -> None:
+    label = _VertexLabel("outer.inner.foo")
+    assert str(label) == "['outer', 'inner', 'foo']"
```
Comment (GeorgeNash) [inline keanu-python/tests/test_labels.py]: what's this test doing?

File: keanu-python/keanu/vertex/label.py [hunk 4]
```diff
@@ -9,6 +10,24 @@
 
 class _VertexLabel(JavaObjectWrapper):
```
Comment (icantsayrural) [inline keanu-python/keanu/vertex/label.py]: We might want to write a comment saying this is for internal use, so people don't try to pass it to `Vertex`.

File: keanu-python/keanu/vertex/label.py [hunk 5]
```diff
@@ -9,6 +10,24 @@
...
         super(_VertexLabel, self).__init__(java_object)
+
+    def get_name(self) -> str:
+        return self.unwrap().getQualifiedName()
+
+    def __repr__(self) -> str:
+        return self.get_name().split(".").__repr__()
+
+    @staticmethod
+    def create_from_list(*names: str) -> '_VertexLabel':
```
Comment (icantsayrural) [inline keanu-python/keanu/vertex/label.py]: What happens when this list is empty?

File: keanu-python/keanu/base.py (hunk 1 above)
Comment (icantsayrural) [reply keanu-python/keanu/base.py]: How does `_val` and `_class` get called by `__getattr__` in the first place? These are actually valid fields of `JavaObjectWrapper` that we set ourselves.

File: keanu-python/tests/test_labels.py (hunk 3 above)
Comment (icantsayrural) [reply keanu-python/tests/test_labels.py]: Checking `__repr__`

File: keanu-python/keanu/base.py (hunk 1 above)
Comment (migwellian) [reply keanu-python/keanu/base.py]: This happens if there's an exception in a constructor of a class that inherits from `JavaObjectWrapper`. In such a case, `_val` and `_class` have not been set.

**Comment by migwellian** [comment]
> Why does `VertexLabel` feel so separate from `Vertex` in python? I expected to see it as a member variable in the `Vertex` class and we don't seem to call any of the methods from `Vertex`?

That's a good observation. We don't need to call it from `Vertex` because:
* for `set_label`, we can pass a string as the label
* for `get_label`, we call a Java method that returns a string.

This is only really used behind the scenes in places like `lambda_model` and (coming soon) `plates`.

File: keanu-python/keanu/vertex/label.py (hunk 4 above)
Comment (migwellian) [reply keanu-python/keanu/vertex/label.py]: I've added a docstring:

        class _VertexLabel(JavaObjectWrapper):
        """
        This class is intended for private use - as a user you should only need to specify string values.
        It's used behind the scenes e.g. in :class:`keanu.vertex.lambda_model.LambdaModel`
        """

File: keanu-python/keanu/vertex/label.py (hunk 5 above)
Comment (migwellian) [reply keanu-python/keanu/vertex/label.py]: good call - I've made it raise an exception in that case

**Pull request merged**